File: 253-meeting-rooms-ii/253-meeting-rooms-ii.py
class Solution:
    def minMeetingRooms(self, intervals: List[List[int]]) -> int:
        intervals = sorted(intervals, key = lambda x: x[0])
        heapRooms = []
        heapq.heappush(heapRooms, intervals[0][1])
        for i in range(1, len(intervals)):
            interval = intervals[i]
            if heapRooms[0] <= interval[0]:
                heapq.heappop(heapRooms)
            heapq.heappush(heapRooms, interval[1])
        return len(heapRooms)

File: 253-meeting-rooms-ii/253-meeting-rooms-ii.java
class Solution {
    public int minMeetingRooms(int[][] intervals) {
        Arrays.sort(intervals, (a, b) -> a[0] - b[0]);
        
        PriorityQueue<Integer> minHeap = new PriorityQueue<>((a, b) -> Integer.compare(a, b));
        
        for (int[] i : intervals) {
            if (minHeap.isEmpty()) {
                minHeap.add(i[1]);
            } else {
                if (minHeap.peek() > i[0]) {
                    minHeap.add(i[1]);
                } else {
                    minHeap.poll();
                    minHeap.add(i[1]);
                }
            }
        }
        return minHeap.size();
    }
}